### DISCUSSION TITLE
New StreamField-based form builder

### DISCUSSION THREAD

**Discussion by BertrandBordage** [wagtail/wagtail#11389] [Ideas]
This upcoming work is funded thanks to [Wagtail’s first hatch](https://www.kickstarter.com/projects/noripyt/wagtails-first-hatch), our first Kickstarter campaign! :tada:

It will be started after most of the work from #4473, most likely in August 2018.

This issue hosts all discussions about the features & API we want in this new form builder.

This work will be based on a form builder I built for a project as a Torchbox subcontractor. See it in action in [one of the videos from the campaign](https://www.kickstarter.com/projects/noripyt/wagtails-first-hatch). Here are the additional features provided by this new form builder compared to the existing one:
- It is based on a `StreamField` instead of an `InlinePanel`, making it more flexible.
- Easily create completely new field types.
- Easily create multiple form builder models with completely different field types in it.
- It supports files, and includes a special image block that generates thumbnails in the admin. All files (including generated thumbnails) are automatically removed when a submission is removed. They are not stored as Wagtail images or documents, but as regular Django files.
- Each form consists in one or multiple steps, that can optionally be named.
- Users can submit a form and modify its content later.
- A Google-Drive-like autosave feature.
- Each submission is stored in relation with a user, or a session object for anonymous users. That way, users can easily resume a form.
- Extendable new admin based on `ModelAdmin`.

What needs to be changed before merge in Wagtail:
- Add the possibility to create forms without steps.
- Add the possibility to choose whether a form can be resumed or not.
- Improve the new admin so it supports at least the same features as the current admin. The new admin has a limited date filter and no CSV export, but adds the possibility of custom filters & custom actions.
- Improve the programmer interface, it started being a nice little API, but unfortunately the project specifications evolved a lot, and some parts are a bit messy. Nothing unorthodox, it’s just not easy to understand, so it needs work.
- Improve the possibilities of the default form fields. For example, the equivalent of `CharField` has no possibility to specify a minimum or maximum length.
- Add tests.
- Add docs.

So there’s already a big chunk of work already done, but we’re not there yet.

Obviously, this new form builder will **not be backwards compatible**. So it will not replace directly the existing form builder, there will be one year or two where both builders will be supported at the same time. My idea is first to make it a third-party app, then integrate it as `wagtail.contrib.stream_forms`, at the same time deprecate `wagtail.contrib.forms`, and in Wagtail 3.0, ditch the old form builder and move `wagtail.contrib.stream_forms` to `wagtail.contrib.forms`.

## Fixed issues

This work will implement solutions for at least these issues:
#1348, #1556, #1564, #3001

**Comment by BertrandBordage** [comment]
This work got its first working version here: https://github.com/noripyt/wagtail-flexible-forms
It’s missing its documentation and still needs some cleaning, as well as tests.

**Comment by johnraz** [comment]
Hello @BertrandBordage,

A few questions here:

* Should https://github.com/noripyt/wagtail-flexible-forms be considered alpha? Or is it stable enough to be used?
* Is there any plans on getting this moving further? Any place where I can follow the roadmap?

Thanks a lot!

**Comment by BertrandBordage** [comment]
@johnraz Yes, please consider it as an alpha.
It is currently being tested in two projects with the current state, and after #4942 is merged I will adjust it to work best with the finally merged version of the React StreamField.
The React StreamField itself will be merged before the 1st of August, according to @gasman.

**Comment by thibaudcolas** [comment]
For anyone interested in this – the StreamField form builder now lives at https://github.com/coderedcorp/wagtail-flexible-forms. All working from the Kickstarter campaign has made its way to Wagtail core with the exception of the form builder.

There are no plans that I’m aware of to get this form builder into core – not to say this can’t / shouldn’t happen, just no one is actively making the case for this or working on this.

**Comment by thibaudcolas** [comment]
Thought I’d share #12287 for anyone interested in this, and [wagtail-form-plugins](https://github.com/laas/wagtail-form-plugins).

**Comment by thibaudcolas** [comment]
👋 proposing to add this to the "Future" area of our roadmap: [RFC 111: v7.3* roadmap updates](https://github.com/wagtail/rfcs/pull/111)

**Reply by thibaudcolas** [reply]
Now on the roadmap: [Flexible forms](https://github.com/wagtail/roadmap/issues/120). I think we could really use a clearer definition of what it is exactly that’s missing with the current forms implementation. It’s pretty basic compared to dedicated software, but I don’t really know which aspects exactly most warrant improvement in the context of Wagtail.

Or if anyone has good references on how other systems do it, ideally CMS-centric ones, please share!

**Comment by thibaudcolas** [comment]
Reviewing this a bit more, I think the move to StreamField for forms would help by allowing us to leverage the new [block settings](https://github.com/wagtail/roadmap/issues/103) feature, in addition to other flexibility and UX improvements.

StreamField would make it easier to introduce many different built-in and custom field types, each with their different configuration fields. Block settings would make it easier to hide away any of those common or type-specific configuration fields. [Conditional form fields in the CMS](https://github.com/wagtail/wagtail/issues/2172) might help a bit too.

## Form builder requirements overview

Looking at form builder requirements not met by the current system, essentially anything that’s somewhat complex field types or sequence logic would be much easier to build with custom reusable "field types as blocks". That could be defined in third-party packages, like advanced StreamField blocks are.

### Solved with StreamForms

Here’s what integrating [wagtail-flexible-forms](https://github.com/coderedcorp/wagtail-flexible-forms) would give us right now:

| Form builder requirements                             |
| ------------------------------------------------------------ |
| **New field types / widgets**                           |
| Phone field (#6611)                    |
| Plain-text address field                                     |
| Address lookup widget                                        |
| Price / currency field with formatting and symbol            |
| File uploads (images / documents)                            |
| Star rating / scale feedback fields                          |
| Likert scale (slider) or table (matrix of questions)         |
| Time picker (hour/minute widget)                             |
| Predefined choice values (from a snippet, or standardized sets like country, age group, etc) |
| **Field configuration**                           |
| Custom error messages |
| “Other” option with free-text follow-up for radio buttons    |
| Min length / max length checks                               |
| Character counters (single line + textarea)                  |
| Autocomplete attribute (#13629)                    |
| Placeholder attribute (#6132) |
| Conditional validation (required/optional) based on other fields                 |
| Regex / custom validation rules configurable by editors      |
| **Form flow**                            |
| Conditional fields based on rules / other fields                   |
| Multi-section forms                   |
| Multi-step forms with page breaks                            |
| Conditional routing based on field values                    |
| Progress indicators (% complete, step count)                 |
| Branching logic                                              |

### Not solved with StreamForms

And for reference, other requirements that I don’t think would be affected by this kind of StreamField rebuild. All/most of those strike me as better implemented outside Wagtail:

| Form builder requirements                             |
| ------------------------------------------------------------ |
| **Data handling & submission metadata**                      |
| Unique record ID capture                                     |
| SSO user ID capture                                          |
| User agent / IP address capture                                           |
| UTM / referrer data capture                                  |
| Editor control over metadata capture                         |
| Avoid double submission ([for example with a redirect](https://docs.wagtail.org/en/stable/advanced_topics/formbuilder_routablepage_redirect.html)) |
| **Email & notification workflows**                           |
| Multiple notification rules                                  |
| Role-based or dynamic recipients                             |
| Templated logic branches in emails                           |
| Editor-configurable confirmation emails                      |
| Richtext emails with field substitution                      |
| CC/BCC controls                                              |
| Per-form sender identity                                     |
| **Backend, reporting & exports**                             |
| Scheduled (daily/weekly) exports                             |
| Export only when new data exists                             |
| Per-form export configuration                                |
| Field-level filtering                                        |
| Saved queries                                                |
| Role-based access to subsets of data                         |

**Reply by thibaudcolas** [reply]
I think the **Field configuration** requirements listed above are well suited to be provided out of the box. So here’s a quick poll attempt -- **please 👍 upvote the specific ones you think would be most beneficial to have built-in**:

**Reply by thibaudcolas** [reply]
Custom error messages

**Reply by thibaudcolas** [reply]
“Other” option with free-text follow-up for radio buttons

**Reply by thibaudcolas** [reply]
Min length / max length checks

**Reply by thibaudcolas** [reply]
Character counters (single line + textarea)

**Reply by thibaudcolas** [reply]
Autocomplete attribute (https://github.com/wagtail/wagtail/issues/13629)

**Reply by thibaudcolas** [reply]
Placeholder attribute (https://github.com/wagtail/wagtail/issues/6132)

**Reply by thibaudcolas** [reply]
Conditional validation (required/optional) based on other fields

**Reply by thibaudcolas** [reply]
Regex / custom validation rules configurable by editors

**Comment by root-sarvesh** [comment]
From going through the form builder code and recent issues, I feel these items would have the biggest practical impact for editors:

• Min/max length checks
• Character counters
• Placeholder & autocomplete support

These seem both widely requested and feasible to implement cleanly once the StreamField-based form builder takes shape.

I’m keen to contribute to this effort as the new form builder takes shape. I’ll keep following the redesign discussions and am ready to help with implementation work when it becomes actionable.

**Reply by thibaudcolas** [reply]
thank you, though our current implementation is so basic that I don’t really think current capabilities are a relevant part of the equation. The above lists are on the basis we would do a big rebuild first.

**Comment by thibaudcolas** [comment]
Sharing one more implementation from @robmoorman ✨ [wagtail-model-forms](https://github.com/vicktornl/wagtail-model-forms)
